src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.Node.cs

```diff
@@ -738,16 +754,23 @@ internal int IndexOf(T item, int index, int count, IEqualityComparer<T> equality
                 Requires.Range(index + count <= this.Count, nameof(count));
 
                 equalityComparer = equalityComparer ?? EqualityComparer<T>.Default;
-                using (var enumerator = new Enumerator(this, startIndex: index, count: count))
+                int length = index + count;
+
+                if (item == null)
                 {
-                    while (enumerator.MoveNext())
+                    for (int i = index; i < length; i++)
                     {
-                        if (equalityComparer.Equals(item, enumerator.Current))
-                        {
-                            return index;
-                        }
-
-                        index++;
+                        if (this[i] == null)
+                            return i;
+                    }
+                }
+                else
+                {
+                    for (int i = index; i < length; i++)
+                    {
+                        var obj = this[i];
```

Comment:
The indexer is O(log n), which makes this O(n log n).  In contrast, using the enumerator to enumerate is only O(n), but the enumerator has higher constant costs.

I don't think we should take this change as is.  If we're going to start using the indexer as part of iteration like this because it proves to overall be the right choice, we should instead change the enumerator's implementation to be based on the indexer.  But that's going to require serious perf analysis.

Reply:
@stephentoub I'm not married to this as it a bit outside the scope of the issue. I origionally tried to speed up Contains through IndexOf which is where these changes came from. If you think the Contains changes are OK but these are not I'm happy to remove.

Reply:
I support a change to replace the Enumerator implementation. I guess its high fixed cost is from its attempt to be alloc free on average even though it needs a `Stack<T>` object. Maybe we could do a size check on the collection and base the enumerator on the indexer for collection sizes of less than 1M or something like that.

Reply:
@stephentoub @AArnott I've removed my changes to `IndexOf`, I'll take a look at doing what you propose in the enumerator seperatly.

Reply:
@stephentoub @AArnott Hey guys, any updates? Do you want me to resolve this comment?

Reply:
Looks good to me.

Reply:
@AArnott finally got around to looking at at the `Enumerator` code, I can see what your saying about investigating if it would be quicker to use the indexer and I think I'll take a look at that. One thing I don't understand is your comment on using the indexes on collections less then 1M, why would the `Stack` implementation perform better in that case?

Reply:
I was trying to say that given the indexer is faster than acquiring a Stack in order to enumerate properly for collections under 1 million in size, we can re-implement the enumerator in terms of the indexer. But we might implement the enumerator to use either of two modes: stack vs. indexer mode and use the stack one for very large collection sizes so that we don't regress perf for large collections.

That said, I don't know how common large immutable collections are.